research/audioset/vggish/vggish_export_tfhub.py

```diff
@@ -0,0 +1,124 @@
+0.96s example generated from the waveform.
+
+Usage:
+  export_tfhub.py <path/to/VGGish/checkpoint> <path/to/tfhub/export>
+"""
+
+import sys
+sys.path.append('..')  # Lets us import yamnet modules from sibling directory.
+
+import numpy as np
+import resampy
+import tensorflow as tf
+assert tf.version.VERSION >= '2.0.0', (
+    'Need at least TF 2.0, you have TF v{}'.format(tf.version.VERSION))
+import tensorflow_hub as tfhub
+
+import vggish_input
+import vggish_params
+import vggish_slim
+from yamnet import features as yamnet_features
+from yamnet import params as yamnet_params
+
+
+def vggish_definer(variables, checkpoint_path):
+  """Defines VGGish with variables tracked and initialized from a checkpoint."""
+  reader = tf.compat.v1.train.NewCheckpointReader(checkpoint_path)
+
+  def var_tracker(next_creator, **kwargs):
+    """Variable creation hook that assigns initial values from a checkpoint."""
+    var_name = kwargs['name']
+    var_value = reader.get_tensor(var_name)
+    kwargs.update({'initial_value': var_value})
+    var = next_creator(**kwargs)
+    variables.append(var)
+    return var
+
+  def waveform_to_features(waveform):
+    """Creates VGGish features using the YAMNet feature extractor."""
+    yamnet_params.SAMPLE_RATE = vggish_params.SAMPLE_RATE
+    yamnet_params.STFT_WINDOW_SECONDS = vggish_params.STFT_WINDOW_LENGTH_SECONDS
+    yamnet_params.STFT_HOP_SECONDS = vggish_params.STFT_HOP_LENGTH_SECONDS
+    yamnet_params.MEL_BANDS = vggish_params.NUM_MEL_BINS
+    yamnet_params.MEL_MIN_HZ = vggish_params.MEL_MIN_HZ
+    yamnet_params.MEL_MAX_HZ = vggish_params.MEL_MAX_HZ
+    yamnet_params.LOG_OFFSET = vggish_params.LOG_OFFSET
+    yamnet_params.PATCH_WINDOW_SECONDS = vggish_params.EXAMPLE_WINDOW_SECONDS
+    yamnet_params.PATCH_HOP_SECONDS = vggish_params.EXAMPLE_HOP_SECONDS
+    log_mel_spectrogram = yamnet_features.waveform_to_log_mel_spectrogram(
+        waveform, yamnet_params)
+    return yamnet_features.spectrogram_to_patches(
+        log_mel_spectrogram, yamnet_params)
+
+  def define_vggish(waveform):
+    with tf.variable_creator_scope(var_tracker):
+      features = waveform_to_features(waveform)
+      return vggish_slim.define_vggish_slim(features, training=False)
+
+  return define_vggish
+
+
+class VGGish(tf.Module):
+  """A TF2 Module wrapper around VGGish."""
+  def __init__(self, checkpoint_path):
+    super().__init__()
+    self._variables = []
+    self._vggish_fn = tf.compat.v1.wrap_function(
+        vggish_definer(self._variables, checkpoint_path),
+        signature=(tf.TensorSpec(shape=[None], dtype=tf.float32),))
+
+  @tf.function(input_signature=(tf.TensorSpec(shape=[None], dtype=tf.float32),))
+  def __call__(self, waveform):
+    return self._vggish_fn(waveform)
+
+
+def check_model(model_fn):
+  """Applies vggish_smoke_test's sanity check to an instance of VGGish."""
+  num_secs = 3
+  freq = 1000
+  sr = 44100
+  t = np.linspace(0, num_secs, int(num_secs * sr))
```

Comment:
np.arange(0, num_secs, 1 / sr) is more readable to me.

Reply:
One difference is that linspace() turns out to include the endpoint by default, while arange() doesn't. I think arange() is better for readability and for providing a half-open interval so switched to arange() everywhere.

Reply:
So in fact linspace wasn't implementing sample rate = sr, but something slightly less.